lib/compiler-singlepass/src/codegen_x64.rs

```diff
@@ -429,56 +432,48 @@ impl<'a> FuncGen<'a> {
                     ),
                     Location::GPR(base_reg),
                 );
-                let current_burnt_reg = self.machine.acquire_temp_gpr().unwrap();
-                // Read current gas counter.
-                self.assembler.emit_mov(
-                    Size::S64,
-                    Location::Memory(base_reg, counter_offset),
-                    Location::GPR(current_burnt_reg),
-                );
-                // Read opcode cost.
-                let count_reg = self.machine.acquire_temp_gpr().unwrap();
-                self.assembler.emit_mov(
-                    Size::S64,
-                    Location::Memory(base_reg, opcode_cost_offset),
-                    Location::GPR(count_reg),
-                );
-                // Multiply instruction count by opcode cost.
-                match count_location {
-                    Location::Imm32(imm) => self.assembler.emit_imul_imm32_gpr64(imm, count_reg),
-                    _ => self.assembler.emit_imul(
-                        Size::S32,
-                        count_location,
-                        Location::GPR(count_reg),
-                    ),
+
+                // Prepare the registers for the `mul` instruction. `MUL SRC` does effectively a
+                //
+                //   RDX:RAX ← RAX ∗ SRC;
+                //
+                // where SRC must be a register or a memory location. We always have
+                // `opcode_cost_offset` in a memory location, so we should use that as the memory
+                // operand.
+                self.assembler.emit_push(Size::S64, Location::GPR(GPR::RDX));
```

Comment:
We could try to optimize this `push`/`pop` out if we can obtain `RDX` as a temporary GPR.